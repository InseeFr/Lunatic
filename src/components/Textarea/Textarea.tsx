import { type ChangeEventHandler } from 'react';
import './Textarea.scss';
import type { LunaticComponentProps } from '../type';
import { Label } from '../shared/Label/Label';
import {
	ComponentErrors,
	getComponentErrors,
} from '../shared/ComponentErrors/ComponentErrors';
import { slottableComponent } from '../shared/HOC/slottableComponent';
import { Declarations } from '../shared/Declarations/Declarations';
import type { LunaticError } from '../../use-lunatic/type';
import { CustomInput } from '../Input/Input';

export function Textarea({
	handleChange,
	response,
	errors,
	...props
}: LunaticComponentProps<'Textarea'>) {
	return (
		<CustomTextarea
			{...props}
			onChange={(v) => handleChange(response, v)}
			errors={getComponentErrors(errors, props.id)}
		/>
	);
}

type CustomProps = Omit<
	LunaticComponentProps<'Textarea'>,
	'response' | 'handleChange' | 'errors'
> & {
	onChange: (v: string) => void;
	errors?: LunaticError[];
};

export const CustomTextarea = slottableComponent<CustomProps>(
	'Textarea',
	(props) => {
		const {
			value,
			onChange,
			disabled,
			required,
			maxLength,
			label,
			description,
			id,
			errors,
			readOnly,
			declarations,
			rows,
			cols,
			placeHolder,
		} = props;
		const labelId = `label-${id}`;

		return (
			<div className="lunatic-textarea">
				<Label htmlFor={id} id={labelId} description={description}>
					{label}
				</Label>
				<Declarations
					type="AFTER_QUESTION_TEXT"
					declarations={declarations}
					id={id}
				/>
				<textarea
					required={required}
					disabled={disabled}
					id={id}
					rows={rows}
					maxLength={maxLength}
					cols={cols}
					onChange={(e) => onChange(e.target.value)}
					value={value ?? ''}
					placeholder={placeHolder}
					readOnly={readOnly}
					aria-invalid={!!errors}
				/>
				<ComponentErrors errors={errors} />
			</div>
		);
	}
);
