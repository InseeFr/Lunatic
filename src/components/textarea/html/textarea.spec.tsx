import { render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import Textarea from './textarea';

describe('Textarea', () => {
	const mockOnChange = vi.fn();

	beforeEach(() => {
		mockOnChange.mockClear();
	});

	it('renders without crashing', () => {
		const container = render(
			<Textarea
				value={'input'}
				id="input"
				aria-labelledby="input"
				onChange={mockOnChange}
			/>
		);
		expect(container).toMatchSnapshot();
	});

	it('renders input readable but not editable when readOnly prop is true', () => {
		const { container } = render(
			<Textarea
				id="textarea"
				value={
					'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'
				}
				readOnly
				onChange={mockOnChange}
			/>
		);
		expect(container).toMatchSnapshot();

		const input = container.querySelector('textarea');
		expect(input).toHaveAttribute('readOnly');
		(input as HTMLElement).focus();
		expect(input).toHaveFocus();
		expect(input).toHaveValue(
			'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'
		);
	});
});
