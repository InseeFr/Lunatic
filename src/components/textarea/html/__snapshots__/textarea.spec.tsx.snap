// Vitest Snapshot v1

exports[`Textarea > renders input readable but not editable when readOnly prop is true 1`] = `
<div>
  <div
    class="lunatic-textarea"
  >
    <textarea
      id="textarea"
      readonly=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
    </textarea>
  </div>
</div>
`;

exports[`Textarea > renders without crashing 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="lunatic-textarea"
      >
        <textarea
          id="input"
        >
          input
        </textarea>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="lunatic-textarea"
    >
      <textarea
        id="input"
      >
        input
      </textarea>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
